Gear-665 Metadata & Gear-658 Config

margaretmahan opened the pull request:
ENH: Upgrade dcm2niix algorithm version to latest.

- The most applicable dcm2niix version upgrade includes fixes for BIDS RepetitionTimeExcitation and RepetitionTimeInversion tags (https://github.com/rordenlab/dcm2niix/issues/439)

ENH: Incorporate additional dcm2niix command options including NRRD format, comments, and verbose logging.
MAIN: Update instance tests accordingly.
DOC: Updates to README regarding metadata captured from this Gear.

Instance tests run on ss.ce using dcm2niix_rewrite_test collection as inputs.

joshicola commented on tests/instance/assess_instance_runs.py:
```diff
@@ -1,25 +1,26 @@
+versions = ["1.0.0_1.0.20200331", "1.1.0_1.0.20201102_dev0"]
+
 
 input_dir = "~/Documents/flywheel/gears/dcm2niix/tests/instance"
```
This makes this test dependent on a valid user's home directory and the existing subdirectories.

I don't see existence tested or creation enforced anywhere.

Not a blocker..... thoughts toward when we have minicore to test on.

margaretmahan replied:
Again, we do not yet have CI/CD (automation) of end-to-end testing of Gears. These instance tests are for my testing of the Gear in place of that automation and are not meant to be "production-ready". I understand that other Gear authors may not do it this way and so that might be confusing. Think of it as a way to automate end-to-end testing in an "as best we can" sort of way right now without actually needing to click buttons in the UI.

joshicola commented on manifest.json:
```diff
@@ -125,7 +131,7 @@
       },
       "filename": {
-          "description": "Output filename template (%a=antenna (coil) number, %c=comments, %d=description, %e=echo number, %f=folder name, %i=ID of patient, %j=seriesInstanceUID, %k=studyInstanceUID, %m=manufacturer, %n=name of patient, %p=protocol, %s=series number, %t=time, %u=acquisition number, %v=vendor, %x=study ID; %z=sequence name). '%f' (default).",
+          "description": "Output filename template. Options: %a=antenna (coil) number, %c=comments, %d=description, %e=echo number, %f=folder name, %i=ID of patient, %j=seriesInstanceUID, %k=studyInstanceUID, %m=manufacturer, %n=name of patient, %p=protocol, %s=series number, %t=time, %u=acquisition number, %v=vendor, %x=study ID, %z=sequence name. Can combine any of the options into a phrase (e. ...
```
This is more a "filename_format" (like date format "%YY-%MM-%DD").

margaretmahan replied:
I agree. But we are trying to maintain as much parity with the dcm2niix tool and with former dcm2niix Gear versions.

joshicola commented on run.py:
```diff
@@ -29,14 +29,14 @@ def main(gear_context):
+        output_image_files = None
 
-    if not isinstance(nifti_files, list):
+    if not isinstance(output_image_files, list):
```
Unless I am missing something, the logic above and the `list` or `str` output of `output.outputs.converted_files` guarantees that `output_image_files` will be either a `list` or a `None`.

If this is so, can this statement be reduced to:

`if not output_image_files:`

I am not familiar enough with the `output` to know what else there could be.

margaretmahan replied:
I had to enforce some typing on the `output` object because it has a variable datatype return status. The `output` object is of type `nipype.interfaces.base.support.InterfaceResult` which can be a string or a list (as commented on line 30). From line 31 to line 37, if there is an output, then it is forced to a list datatype. This is because you can have one or many files in the output (think, diffusion has .bvecs and .bvals, but functional may only have one .nii) but we want to maintain the same datatype going forward. Hence, forcing the list datatype and defaulting to None, if not possible. From line 39 to line 51, it is really just ignoring the previous lines based on a config option. But I agree, there's a bit of a logic nightmare going on here. Some from the variable datatype return status of output object and some from config options. Something to add to the TODO list and think about how to make it cleaner on the next update round.

joshicola commented on dcm2niix/dcm2niix_run.py:
```diff
@@ -16,18 +16,21 @@ def convert_directory(
     convert_only_series="all",
     crop=False,
-    filename="%f",
+    filename="%f_%p_%t_%s",
```
Your `manifest.json` lists the default as "%f".

Did you want consistency across these defaults?

margaretmahan replied:
Here's my thinking: the scripts that perform the actions, like this one is to run the dcm2niix tool, should have the defaults set to the tool's defaults. So if you were to pip install this `dcm2niix` into another Gear, you would have all the defaults as implied by the dcm2niix tool. In this case, the dcm2niix tool defaults to `filename="%f_%p_%t_%s"`, which is why I changed it to this. But the Gear's defaults (via legacy) is `filename="%f"`. I did add more to the `README.md` to clarify this. And I am open to changing it back if you think that would be best.

joshicola commented on dcm2niix/dcm2niix_run.py:
```diff
@@ -126,6 +133,37 @@ def convert_directory(
             # See: https://www.nitrc.org/forum/forum.php?thread_id=11134&forum_id=4703
             converter.inputs.series_numbers = convert_only_series
 
+        # dcm2niix command configurations for: crop
```
I like how you expanded this section to include comments for the dcm2niix converter options.

margaretmahan replied:
:+1:

joshicola reviewed (approved):
Thank you for the opportunity to review.

It is always illuminating to see how others structure their thoughts in "gear form".

There are no blockers. 

I have some general questions and comments around clarity and form.

Otherwise, I like what you have done with the place... I meant "Pydeface".

margaretmahan commented on manifest.json:
```diff
@@ -12,7 +12,7 @@
        "gear-builder": {
            "category": "converter",
-           "image": "flywheel/dcm2niix:1.0.0_1.0.20200331"
+           "image": "flywheel/dcm2niix:1.1.0_1.0.20201102_dev0"
```
I will update the version just before merging with master.

margaretmahan commented on utils/resolve.py:
```diff
@@ -74,47 +79,53 @@ def setup(
 
```
It depends on the grouping of the concept and I use spacing to make it easier to read the code when there are multiple flow control statements.

margaretmahan commented on utils/resolve.py:
```diff
@@ -124,15 +135,23 @@ def retain_gear_outputs(
+        )
+        os.sys.exit(1)
+
     # fmt: off
```
The purpose of a code formatter is to make it easier to read and consistent across the codebase. I make sure to double-check between the `#fmt: off` and `#fmt: on` for consistency with black. However, if you run black on this script, it will drastically change the structure and spacing, making it extremely difficult to read and debug. If you notice, there are several sections that are very similar in nature, and maintaining the consistency of structure across the script is more important for longevity and maintenance. In addition, this is the only place, less than 100 lines of code, in this entire Gear where I have elected to turn the black formatter off, so I do not believe this is an issue worth attention. As such, I will be retaining the `#fmt off/on` options in this script.

margaretmahan commented on utils/resolve.py:
```diff
@@ -124,15 +135,23 @@ def retain_gear_outputs(
             bids_sidecar = os.path.join(
                                         work_dir, re.sub(
-                                                         r"(\.nii\.gz|\.nii)",
+                                                         r"(\.nii\.gz|\.nii|.nhdr|\.raw\.gz)",
```
The `.nhdr` is not necessary but retained for clarity and consistency. If the config option is NIfTI files, then `\.nii\.gz|\.nii` is necessary. If config option is NRRD, then `.nhdr|\.raw\.gz` is necessary. The distinction between these two options is upstream and so they are both retained here as I do not see a reason to pull that out as a variable and add complexity. I've added a note to try and refactor this script on the next maintenance round.

margaretmahan commented on utils/resolve.py:
```diff
@@ -74,47 +79,53 @@ def setup(
 
 def retain_gear_outputs(
-    nifti_files,
+    output_image_files,
```
Okay.

margaretmahan commented on utils/parse_config.py:
```diff
@@ -64,22 +64,35 @@ def generate_gear_args(gear_context, FLAG):
-        # user-defined config option setting wil be considered during gear resolve.
+        # user-defined config option setting wil be considered during the gear resolve stage.
+        filename = gear_context.config["filename"]
+        filename = filename.replace(" ", "_")
```
The dcm2niix tool will catch errors related to the filename. And errors that impact the output files in Flywheel would then be caught on Gear completion. I think it would be good to have the sanitization of output filenames, but I am not sure all that that entails and is really more of a Gear construction issue than a specific dcm2niix issue - or rather something for the Gear Toolkit. Something to bring to the team at least!

margaretmahan commented on utils/metadata.py:
```diff
@@ -31,13 +32,16 @@ def generate(
 
     Args:
-        nifti_files (list): The absolute paths to nifti files to resolve.
+        output_image_files (list): The absolute paths to gear image files to resolve.
```
Not sure what you are suggesting here.

margaretmahan commented on utils/metadata.py:
```diff
@@ -50,16 +54,22 @@ def generate(
     """
     log.info("Generating metadata.")
 
+    if retain_nifti == output_nrrd:
```
And it does. On line 160 of `parse_config.py`, I swap the `retain_nifti` config to `False` if `output_nrrd` is `True`, because that's the expected behavior of dcm2niix.

I also place that logic here because it is relevant to the parameters of this function in particular. Future utilization then only needs to look at this function to be understood as the logic and functionality are explicitly retained into one unit.

margaretmahan commented on utils/metadata.py:
```diff
@@ -135,7 +145,9 @@ def generate(
         # Apply collated metadata to all associated files
-        if retain_sidecar:
+
+        # Sidecar; if NRRD format, two files per sidecar - capture sidecar once
```
That might be true but that is not the output of the dcm2niix, which produces the header as a separate file (at least in everything I tested). I am not familiar with the NRRD file format so if you think the logic could be broken, please do share more!

margaretmahan commented on utils/metadata.py:
```diff
@@ -50,16 +54,22 @@ def generate(
+
     capture_metadata = []
 
     # fmt: off
```
See previous comment regarding black formatting.

margaretmahan commented on tests/instance/assess_instance_runs.py:
```diff
@@ -1,25 +1,26 @@
 import flywheel
 import pandas as pd
 
+versions = ["1.0.0_1.0.20200331", "1.1.0_1.0.20201102_dev0"]
```
I am only placing these scripts with the repo to keep them together and not lose anything. As you know, instance tests are not yet automated because we do not yet have a solution for end-to-end testing of Gears. Therefore, these scripts are really just for my testing of the Gear, they are not meant to be "production-ready".

margaretmahan commented:
Thanks @joshicola for your review! I think I addressed all your concerns and have made TODOs on a couple of thoughts you had to try and clean up some things on the next maintenance round.
